### PR TITLE
Feat/file type hints collector

### LINKED ISSUE #98: Collect type hint information of files
Collect the type hint information of any given python file by generating an AST & reading the type hints of the nodes. Return a dataframe containing the information.

### PR DESCRIPTION
**Pull request by RecurvedBow**
Closes #98 , #103

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -54,3 +56,21 @@ class TraceData:
...
+        VARTYPE: pd.StringDtype(),
+    }
+
+    METRICS_SCHEMA = {
+        FILENAME: pd.StringDtype(),
+        CLASS: pd.StringDtype(),
+        CATEGORY: pd.Int64Dtype(),
+        COMPLETENESS: pd.BooleanDtype(),
+        CORRECTNESS: pd.BooleanDtype(),
+    }
```
Comment (bengsparks) [inline constants.py:76]: I think we should store the compared symbols in here too, including the module path and the type name from BOTH sides, so that it is easier to see what differs

File: tests/typegen/evaluation/test_file_type_hints_collector.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+def test_file_type_hints_collector_returns_correct_data():
+    expected_trace_data_file_path = path_sample_data_file / "sample1.pytype"
+    expected_typehint_data = pd.read_pickle(expected_trace_data_file_path)
+    expected_typehint_data = expected_typehint_data.astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+    test_object = FileTypeHintsCollector(pathlib.Path.cwd())
+    test_object.collect_data([path_sample_file])
+    actual_typehint_data = test_object.typehint_data
+
+    with pd.option_context("display.max_rows", None, "display.max_columns", None):
```
Comment (bengsparks) [inline tests/typegen/evaluation/test_file_type_hints_collector.py:19]: This `option_context` is no longer needed, I set it in the `tests/conftest.py`

File: tests/typegen/evaluation/test_file_type_hints_collector.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+def test_file_type_hints_collector_returns_correct_data_for_multiple_files():
+    expected_trace_data_file_path = path_sample_data_file / "sample2.pytype"
+    expected_typehint_data = pd.read_pickle(expected_trace_data_file_path)
+    expected_typehint_data = expected_typehint_data.astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+    test_object = FileTypeHintsCollector(pathlib.Path.cwd())
+    test_object.collect_data([path_sample_file, path_sample_file])
+    actual_typehint_data = test_object.typehint_data
+    with pd.option_context("display.max_rows", None, "display.max_columns", None):
```
Comment (bengsparks) [inline tests/typegen/evaluation/test_file_type_hints_collector.py:34]: Here too

File: typegen/evaluation/file_type_hints_collector.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import constants
+from tracing import TraceDataCategory
+
+
+class FileTypeHintsCollector:
+    """Collects the type hints of multiple .py files."""
+    def __init__(self, project_dir: pathlib.Path):
+        self.project_dir = project_dir
+        self.typehint_data = pd.DataFrame()
+        self.file_path = ""
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:16]: Is this unused?

File: typegen/evaluation/file_type_hints_collector.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    def collect_data(self, file_paths: list[pathlib.Path]) -> None:
+        """Collects the type hints of the provided file paths."""
+        for file_path in file_paths:
+            with file_path.open() as file:
+                file_content = file.read()
+
+            module = cst.parse_module(source=file_content)
+            module_and_meta = cst.MetadataWrapper(module)
+            relative_path = str(pathlib.Path(os.path.relpath(file_path.resolve(), self.project_dir.resolve())))
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:26]: Use the `relative_to` method on `pathlib.Path` instances

File: typegen/evaluation/file_type_hints_collector.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+            self.typehint_data = pd.concat(
+                [self.typehint_data, typehint_data], ignore_index=True
+            ).astype(constants.TraceData.TYPE_HINT_SCHEMA)
+
+
+class _TypeHintVisitor(cst.CSTVisitor):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+
+    def __init__(self, file_path: str) -> None:
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:39]: Which would make this a `pathlib.Path` too

File: typegen/evaluation/file_type_hints_collector.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        variable_name = None if not hasattr(node, "name") else node.name.value
+        variable_line_number = pos.line
+        return variable_name, variable_line_number
+
+    def _add_row(self, line_number: int, category: TraceDataCategory, variable_name: str | None, type_hint: str | None):
+        class_node = self._innermost_class()
+        class_name = None
+        if class_node:
+            class_name = class_node.name.value
+        function_node = self._innermost_class()
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:129]: Looks like a C&P error. This should probably be `self._innermost_function`. Update tests to reflect this bug

File: typegen/evaluation/file_type_hints_collector.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+            variable_name = node.target.attr.value
+        elif isinstance(node.target, cst.Name):
+            category = TraceDataCategory.LOCAL_VARIABLE
+            variable_name = node.target.value
+        else:
+            raise TypeError("Unhandled case for: " + type(node.annotation.annotation).__name__)
+        self._add_row(line_number, category, variable_name, type_hint)
+        return True
+
+    def _get_name_and_line_number(self, node):
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:118]: type hints missing here

File: typegen/evaluation/file_type_hints_collector.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            type_hint,
+        ]
+
+    def _get_annotation_value(self, annotation: cst.Annotation | None) -> str | None:
+        if annotation is None:
+            return None
+
+        actual_annotation = annotation.annotation
+        if isinstance(actual_annotation, cst.Attribute):
+            module_name = actual_annotation.value.value  # type: ignore
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:149]: Shouldn't have to `type: ignore` here, this is probably hiding API misusage

File: typegen/evaluation/file_type_hints_collector.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            self.file_path,
+            class_name,
+            function_name,
+            line_number,
+            category,
+            variable_name,
+            type_hint,
+        ]
+
+    def _get_annotation_value(self, annotation: cst.Annotation | None) -> str | None:
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:143]: This does not cover the case
```
from x import y

def f(inst: y.z): pass
```

Update tests to check this bug (if I wrong, then it is a good test to have at least :) )

File: constants.py (hunk 1 above)
Comment (bengsparks) [reply constants.py:76]: What is the symbol name in this schema?

e.g. if I have `x: int = 5`, where are `x` and `int` stored?

File: typegen/evaluation/metric_data_calculator.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import numpy as np
+import pandas as pd
+
+import constants
+
+
+class MetricDataCalculator:
+    """Calculates the metric data containing the correctness & completeness."""
+    def __init__(self):
+        self.modified_filenames_by_original: Dict[str, str] = {}
```
Comment (bengsparks) [inline typegen/evaluation/metric_data_calculator.py:12]: minor nitpick: prefer `dict` over its `Dict` counterpart

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: typegen/evaluation/metric_data_calculator.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        subset_correctness = list(constants.TraceData.TYPE_HINT_SCHEMA.keys())
+        subset_completeness = subset_correctness.copy()
+        subset_completeness.remove(constants.TraceData.VARTYPE)
+
+        # Calculates the completeness by checking
+        # which rows of the original data are contained in the traced data (without type comparison).
+        traced_type_hint_data_without_types = modified_type_hint_data.drop(constants.TraceData.VARTYPE, axis=1)
+        merged_data_completeness = pd.merge(original_type_hint_data, traced_type_hint_data_without_types,
+                                            on=subset_completeness, how='left',
+                                            indicator=constants.TraceData.COMPLETENESS)  # type: ignore
```
Comment (bengsparks) [inline typegen/evaluation/metric_data_calculator.py:34]: What is being ignored here?

File: typegen/evaluation/metric_data_calculator.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        merged_data_completeness[constants.TraceData.COMPLETENESS] = \
+            np.where(merged_data_completeness[constants.TraceData.COMPLETENESS] == "both", True, False)
+
+        # Calculates the correctness by checking
+        # which rows of the original data are contained in the traced data (with type comparison).
+        merged_data_completeness_correctness = pd.merge(merged_data_completeness, modified_type_hint_data,
+                                                        on=subset_correctness, how='left',
+                                                        indicator=constants.TraceData.CORRECTNESS)  # type: ignore
```
Comment (bengsparks) [inline typegen/evaluation/metric_data_calculator.py:42]: And here?

File: typegen/evaluation/file_type_hints_collector.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+    def _innermost_function(self) -> cst.FunctionDef | None:
+        fromtop = reversed(self._scope_stack)
+        fdefs = filter(lambda p: isinstance(p, cst.FunctionDef), fromtop)
+
+        first: cst.FunctionDef | None = next(fdefs, None)  # type: ignore
+        return first
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> bool | None:
+        module_name: str = str(node.module.value)  # type: ignore
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:61]: This probably shouldn't be ignored. If the type is there, that means there is another case to cover

File: typegen/evaluation/file_type_hints_collector.py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+        for name in node.names:  # type: ignore
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:62]: See below: the case `from x import y` and `y` is not a class is not covered here.
But try it in a test, if I am wrong, then this is OK

File: typegen/evaluation/file_type_hints_collector.py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+            class_name: str = str(name.name.value)
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:63]: No conversion should be necessary. value is a `str`

File: typegen/evaluation/file_type_hints_collector.py [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    def _add_row(self, line_number: int, category: TraceDataCategory, variable_name: str | None, type_hint: str | None):
+        class_node = self._innermost_class()
+        class_name = None
+        if class_node:
+            class_name = class_node.name.value
+        function_node = self._innermost_class()
+        function_name = None
+        if function_node:
+            function_name = function_node.name.value
+        self.typehint_data.loc[len(self.typehint_data.index)] = [
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:133]: I know we do this in our test data, but this kind of insertion should never be done in code that actually "runs"

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: constants.py (hunk 1 above)
Comment (RecurvedBow) [reply constants.py:76]: The metrics schema contains the columns for the metric data. The metric data is used to determine the correctness & completeness. I think you're thinking of the Type Hint Data.

File: typegen/evaluation/file_type_hints_collector.py (hunk 5 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:26]: The relative_to does not work for example on "a/b/c/d" and "a/b/e", while the os path relpath does.

File: typegen/evaluation/metric_data_calculator.py (hunk 12 above)
Comment (RecurvedBow) [reply typegen/evaluation/metric_data_calculator.py:34]: indicator expects a boolean. But str does work.

File: typegen/evaluation/metric_data_calculator.py (hunk 13 above)
Comment (RecurvedBow) [reply typegen/evaluation/metric_data_calculator.py:42]: Same as above

File: typegen/evaluation/file_type_hints_collector.py (hunk 17 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:133]: Why not? It is only one row.

File: typegen/evaluation/file_type_hints_collector.py (hunk 6 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:39]: There isn't any usage for file path being pathlib.Path. In fact, if it were a path, it would have to be converted to a str.

File: tests/typegen/evaluation/test_file_type_hints_collector.py (hunk 2 above)
Comment (RecurvedBow) [reply tests/typegen/evaluation/test_file_type_hints_collector.py:19]: Done

File: tests/typegen/evaluation/test_file_type_hints_collector.py (hunk 3 above)
Comment (RecurvedBow) [reply tests/typegen/evaluation/test_file_type_hints_collector.py:34]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 4 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:16]: Removed

File: constants.py (hunk 1 above)
Comment (bengsparks) [reply constants.py:76]: Yes, but how do you associate a metric with the compared symbol?

File: typegen/evaluation/file_type_hints_collector.py (hunk 5 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:26]: that's because those paths are not relative to each other.
But you are only loading file from within the project directory, as these are the files we want to type hint, so this case cannot occur. If it does, that means a bug is somewhere else.

File: typegen/evaluation/file_type_hints_collector.py (hunk 6 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:39]: Ah I see, you only use it to write into the DataFrame? That's fine then :+1:

File: typegen/evaluation/file_type_hints_collector.py (hunk 17 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:133]: Because it could overwrite data that already exists at the index. You want to `append` to the DataFrame

File: typegen/evaluation/file_type_hints_collector.py (hunk 15 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:62]: Bonus: all imports relevant for type hinting are in a `if TYPE_CHECKING:` block close to the top of the file; only consider these imports instead of all imports the files has.

File: constants.py (hunk 1 above)
Comment (bengsparks) [reply constants.py:76]: You could perhaps store the index of the compared symbol in the Metrics DataFrame as a kind of foreign key into the parsed AST DataFrame

File: typegen/evaluation/file_type_hints_collector.py (hunk 15 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:62]: As per yesterday's discord discussion:

Our generated file will only contain type hints that have imports from the  `if TYPE_CHECKING:` block. 
Therefore, all type hints in the generated file must be connected to their importing statement from within this block. What I am worried about is that an already existing import in the file contains a symbol by the same name that would make the imported class ambiguous.
The fix for this is very simple, and could preemptively save us a lot of bug-tracking. 
The potential reward to effort ratio is too high to not consider doing this.

File: typegen/evaluation/file_type_hints_collector.py (hunk 5 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:26]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 7 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:129]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 8 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:118]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 9 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:149]: Fixed

File: typegen/evaluation/file_type_hints_collector.py (hunk 10 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:143]: Added

File: typegen/evaluation/metric_data_calculator.py (hunk 11 above)
Comment (RecurvedBow) [reply typegen/evaluation/metric_data_calculator.py:12]: Fixed

File: typegen/evaluation/file_type_hints_collector.py (hunk 14 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:61]: Fixed

File: typegen/evaluation/file_type_hints_collector.py (hunk 16 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:63]: Fixed

File: typegen/evaluation/file_type_hints_collector.py (hunk 17 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:133]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 15 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py]: There is no such constraint that files contain this if TYPE_CHECKING block. Additionally, all imports are important. They provide the information of the module name of a class name. It is not possible to skip any import.

However, the issue of ambiguity might be a valid bug. I will test it.

File: constants.py (hunk 1 above)
Comment (RecurvedBow) [reply constants.py:76]: After some thinking, the correctness & completeness is used to compare filter lists. However, the values of the columns are not causes for changes in correctness & completeness. Thus, causal inference should not occur. Due to that, only the correctness & completeness metric are computed and returned.

Comment (RecurvedBow) [reply constants.py]: Now, the overlap between two type hint data is considered as the metric data.

File: typegen/evaluation/file_type_hints_collector.py (hunk 16 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py]: Then add an `assert isinstance` statement, so that we get a clear error traceback if something did go wrong

File: typegen/evaluation/file_type_hints_collector.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+        return first
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> bool | None:
+        module_name = self._get_module_name(node)
+        if not isinstance(node.names, cst.ImportStar):
+            iterator = iter(node.names)
+            for name in iterator:
+                imported_element_name: str = str(name.name.value)
+                if imported_element_name in self.imports.keys():
+                    # The first import is considered as the module of the imported element.
```
Comment (bengsparks) [inline typegen/evaluation/file_type_hints_collector.py:70]: The last import determines the correct module. You can used `reversed` on `node.names` to fix this.

File: constants.py [hunk 19]
```diff
@@ -24,6 +24,9 @@ class TraceData:
     VARNAME = "VarName"
     VARTYPE_MODULE = "TypeModule"
     VARTYPE = "Type"
+    VARTYPE2 = "Type 2"
```
Comment (bengsparks) [inline constants.py:27]: Perhaps a better name? First one is "original", second is "generated"?

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: constants.py (hunk 1 above)
Comment (RecurvedBow) [reply constants.py]: Since line numbers might not match, indentation level is used instead.

File: typegen/evaluation/file_type_hints_collector.py (hunk 18 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:70]: It is not about determining the module within the names, but rather skipping the imported element if an element with the same name has already been imported.

For example:
```Python
from tests.resource.typegen.evaluation.sample_to_import import FileTypeHintsCollector
from typegen.evaluation import FileTypeHintsCollector
```

And we use FileTypeHintsCollector, then it refers to tests.resource.typegen.evaluation.sample_to_import.FileTypeHintsCollector

The names are iterated to assign the module for each imported element, for example:
```Python
from pathlib import PosixPath, WindowsPath
```

File: constants.py (hunk 19 above)
Comment (RecurvedBow) [reply constants.py:27]: I can't change the name of the first VARTYPE as it is also used in Trace Data & Type Hint Data. I also can't change the name of VARTYPE2 to Generated Type as the  VARTYPE as that is used in Trace Data & Type Hint Data corresponds to the generated type (just not for the metrics data). 

However, if you find more fitting names, I will consider changing the names.

File: typegen/evaluation/file_type_hints_collector.py (hunk 18 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py:70]: It is not about determining the module within the names, but rather skipping the imported element if an element with the same name has already been imported.

For example:
```Python
from tests.resource.typegen.evaluation.sample_to_import import FileTypeHintsCollector
from typegen.evaluation import FileTypeHintsCollector
```

And we use FileTypeHintsCollector, then it refers to tests.resource.typegen.evaluation.sample_to_import.FileTypeHintsCollector

The loop is there to assign the module for each imported element, for example:
```Python
from pathlib import PosixPath, WindowsPath
```

Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py:70]: Ok, tbf I dont think the scenario can occur as this is a symbol conflict.

File: constants.py (hunk 19 above)
Comment (bengsparks) [reply constants.py:27]: Then create two new variables, so that you can name them freely without being dependent on the names of the DataFrames for tracing

File: typegen/evaluation/file_type_hints_collector.py (hunk 16 above)
Comment (RecurvedBow) [reply typegen/evaluation/file_type_hints_collector.py]: Done

File: constants.py (hunk 19 above)
Comment (RecurvedBow) [reply constants.py]: Done

File: typegen/evaluation/file_type_hints_collector.py (hunk 16 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py]: Ah, I have found the reason.
`value` is always a `str`, but `module` can be either an `cst.Attribute`, or a `cst.Name`.
`cst.Attribute` is likely when the case `from x.y import z` occurs; `x.y` being the attribute. This must be tested and fixed prior to merging

File: tests/typegen/evaluation/test_metric_data_calculator.py [hunk 20]
```diff
@@ -10,7 +10,7 @@ def get_sample_data():
         "sample_original_filename",
         None,
         "sample_function_name",
-        10,
+        0,
```
Comment (bengsparks) [inline tests/typegen/evaluation/test_metric_data_calculator.py:13]: Why the change? Is this a relative line number offset from the original function declaration?

File: typegen/evaluation/metric_data_calculator.py [hunk 21]
```diff
@@ -7,40 +7,93 @@
...
 
 class MetricDataCalculator:
+    INDEX = "Index"
+    INDEX2 = "Index 2"
+    CUMCOUNT = "CumCount"
+
     """Calculates the metric data containing the correctness & completeness."""
     def __init__(self):
-        self.modified_filenames_by_original: Dict[str, str] = {}
+        self.generated_filenames_by_original: Dict[str, str] = {}
```
Comment (bengsparks) [inline typegen/evaluation/metric_data_calculator.py:16]: There is still `dict` here :wink:

**Review by bengsparks — CHANGES_REQUESTED** [review]

File: typegen/evaluation/file_type_hints_collector.py (hunk 16 above)
Comment (bengsparks) [reply typegen/evaluation/file_type_hints_collector.py]: Ah, this was fixed outside of the displayed diff. closing

File: tests/typegen/evaluation/test_metric_data_calculator.py (hunk 20 above)
Comment (RecurvedBow) [reply tests/typegen/evaluation/test_metric_data_calculator.py:13]: Instead of using the line number, we now use the column offset. However, even without the column offset, we can still determine the parameter as there can not be another parameter with the same name. Additionally, keeping the column offset causes the following issues:
```Python
def f(a: int, b: bool):
    pass

def f(a: float, b: bool):
    pass
```
The column offsets of the 2 b parameters are different and we do not gain anything by keeping the values. Thus, we can set it to 0.

Comment (bengsparks) [reply tests/typegen/evaluation/test_metric_data_calculator.py:13]: Do we gain anything by using the column data then?

Comment (RecurvedBow) [reply tests/typegen/evaluation/test_metric_data_calculator.py:13]: It is useful for local variables. For example:

```Python
e: int = 1
if True:
    e: str = "string"
```
We can differentiate between the first e and the second e. I have already added this case in the metric data calculator test.

**Review by bengsparks — APPROVED** [review]
